### PR TITLE
fix(ssa): Handle array arguments to side effectual constrain statements

### PR DESCRIPTION
**Pull request by vezenovm**
…ffects

# Description

## Problem\*

Resolves the error that occurs in this PR (https://github.com/noir-lang/noir/pull/3739)

No issue as was discovered and fixed quickly.

## Summary\*

Using an array as an argument in a side effectual constrain statement will cause this panic:
```
Message:  internal error: entered unreachable code: Can only cast to a numeric
Location: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1652
```

Here is an example program which would previously fail that is included as a regression test:
```
struct Bar {
    inner: [u32; 3],
}

fn main(y: pub u32) {
    let bar = Bar { inner: [100, 101, 102] };

    // The assert inside the if should be hit
    if y < 10 {
        assert(bar.inner ==  [100, 101, 102]);
    }

    // The assert inside the if should not be hit
    if y > 10 {
        assert(bar.inner == [0, 1, 2]);
    }
}
```

Without this fix it is not possible to constrain arrays inside of if statements based upon witness conditions.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs [hunk 1]
```diff
@@ -685,6 +672,90 @@ impl<'f> Context<'f> {
...
+                        .function
+                        .dfg
+                        .insert_instruction_and_results(
+                            get,
+                            self.inserter.function.entry_block(),
+                            typevars,
+                            CallStack::new(),
+                        )
+                        .first()
+                };
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs:746]: We have a `insert_array_get` method which I think could make this cleaner. You may need to switch to the entry block and back again if we're not already on it though.

Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs:746]: `insert_array_get` is part of the `FunctionBuilder` while we are modifying the DFG directly here. Unless I am missing something?

**Comment by jfecher** [comment]
Would we still need this fix if assertions just took a boolean in SSA and the equality was a separate instruction?

**Comment by vezenovm** [comment]
> Would we still need this fix if assertions just took a boolean in SSA and the equality was a separate instruction?

I don't think so, as we would only be constraining on the boolean. I went with this solution for now as this seems like an easy bug to run into and making equality a separate instruction is a larger change.

**Review by jfecher — APPROVED** [review]
LGTM

**Pull request merged**